File: day_5/solution.py
import re
from typing import List

def getData(fName):
    """
    @return crates object, instructions list
    """
    with open(fName, "r") as dataFile:
        data = [ i for i in dataFile.readlines() ]
    
    # Crates stored in cols
    crates = {}

    # Load crates
    i = 0
    while "[" in data[i]: # (HACKY) Is a crate

        _crates = re.findall("[A-Z]", data[i])
        # Get each index in string and calculate col
        for cr in _crates:
            cr_i = data[i].index(cr)    # Each crate index in the string
            col = ((cr_i - 1)//4) + 1   # Calculate the col based on the index

            # Load but prepend since we are scanning from top to bottom
            col_crates: List[str] = crates.get(str(col), [])
            col_crates = [cr] + col_crates 
            crates[str(col)] = col_crates

            # Remove the crate from the string
            data[i] = data[i][:cr_i] + " " + data[i][cr_i+1:]

        i += 1

    # Skip the next 2 lines
    i += 2

    instructions: List[List[int]] = []
    # Get Instructions
    while i < len(data):       
        # Extract the parameters
        matches = []
        strBuf = ""        
        j = 0 # 0 means no, 1 means frm, 2 means to
        while j < len(data[i]):
            if data[i][j].isnumeric():
                strBuf += data[i][j]
            elif len(strBuf) > 0:
                # Clear buffer
                matches.append(int(strBuf))
                strBuf = ""

            j += 1

        # Clear buffer again
        if len(strBuf) > 0:
            matches.append(int(strBuf))
            strBuf = ""

        instructions.append(matches)

        i+=1

    return crates, instructions


def part1():
    crates, instructions = getData("data.in")

    for i in instructions:
        # Number of crates, from coloum, to column
        no, frm, to = i 
        frm = str(frm)
        to = str(to)

        # Move
        for _ in range(no):
            crt = crates[frm].pop()
            crates[to].append(crt)

    ans = ""

    for k in range(1, len(crates.keys())+1):
        ans += crates[str(k)][len(crates[str(k)])-1]

    print(ans)


def part2():
    crates, instructions = getData("data.in")

    for i in instructions:
        # Number of crates, from coloum, to column
        no, frm, to = i 
        frm = str(frm)
        to = str(to)

        # Standard copy and paste will do
        removed = crates[frm][len(crates[frm])-no:]
        crates[frm] = crates[frm][:len(crates[frm])-no]
        crates[to] += removed

    ans = ""

    for k in range(1, len(crates.keys())+1):
        ans += crates[str(k)][len(crates[str(k)])-1]

    print(ans)


if __name__ == "__main__":
    print("Part 1 -> ")
    part1()
    print("Part 2 -> ")
    part2()

File: day_5/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <stack>
#include <utility>
#include <map>
using namespace std;

// Define types
#define CRATE_TYPE map<int, stack<char>> 
#define INSTRUCTION_TYPE vector<vector<int>>

// Parse inputs
pair<CRATE_TYPE, INSTRUCTION_TYPE> parseInput(string fName = "data.in") {
    CRATE_TYPE crates;
    INSTRUCTION_TYPE instructions;

    // Open file
    ifstream dataFile{fName};

    string line;
    // Read and parse each line
    while (dataFile) {
        getline(dataFile, line);

        if (line[1] == '1') break; // End of crates
        
        for (int c = 0; c < line.length(); ++c) {
            if (line[c] < 'A' || line[c] > 'Z') continue;
            
            int col = ((c-1)/4) + 1;
            
            crates[col].push((char)line[c]);
        };
    }

    // We need to reverse each stack
    for (auto &a : crates) {
        vector<char> tmp;
        while(!a.second.empty()) {
            tmp.push_back(a.second.top());
            a.second.pop();
        };
        for (auto i : tmp) {
            a.second.push(i);
        };
    }

    // Get instructions
    while (!dataFile.eof()) {
        getline(dataFile, line);
        if (line == "") continue;

        // Parse only numbers
        stringstream ss(line);
        string tmp;
        vector<int> inst;
        int n;
        while (ss >> tmp) {
            try {
                n = stoi(tmp);
            } catch (...) {
                continue;
            };
            inst.push_back(n);
        };
        // Load instructions
        instructions.push_back(inst);
    }

    return {crates, instructions};
};

void solve(int part=1) {
    auto v = parseInput();
    CRATE_TYPE crates = v.first;
    INSTRUCTION_TYPE instructions = v.second;


    // Go through each instructions
    for (auto i : instructions) {
        int no = i[0], frm = i[1], to = i[2];

        // Solve for each parts
        if (part == 1) {
            for (int _ = 0; _ < no; ++_) {
                char tmp = crates[frm].top();
                crates[frm].pop();
                crates[to].push(tmp);
            }
        } else if (part == 2) {
            vector<char> tmpList;
            for (int _ = 0; _ < no; ++_) {
                char tmp = crates[frm].top();
                crates[frm].pop();
                tmpList.insert(tmpList.begin(), tmp);
            }
            for (int crt : tmpList) {
                crates[to].push(crt);
            }
        }
    };

    string ans;

    for (auto x: crates) {
        ans += x.second.top();
    }

    cout << "Part " << part << " = " << ans << endl;
}

int main() {
    solve(1);
    solve(2);

    return 0;
}